messaging/messaging.go

```diff
@@ -0,0 +1,460 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package messaging contains functions for sending messages and managing
+// device subscriptions with Firebase Cloud Messaging (FCM).
+package messaging
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"firebase.google.com/go/internal"
+	"google.golang.org/api/transport"
+)
+
+const (
+	messagingEndpoint = "https://fcm.googleapis.com/v1"
+	iidEndpoint       = "https://iid.googleapis.com"
+	iidSubscribe      = "iid/v1:batchAdd"
+	iidUnsubscribe    = "iid/v1:batchRemove"
+)
+
+var (
+	topicNamePattern = regexp.MustCompile("^(/topics/)?(private/)?[a-zA-Z0-9-_.~%]+$")
+
+	fcmErrorCodes = map[string]string{
```

Comment:
These are mostly the standard Google error codes. Can we make one map that we share across all the SDKs for the standard error codes?

Reply:
We seem to have a few services with non-standard error codes. Also the interpretation of some error codes seem to be different from service to service (e.g. `NOT_FOUND` means one thing in IID, and another in FCM). Lets have a separate discussion about this error code reconciliation.

Reply:
Error codes enumerated in this map does not match those mentioned in the [docs](https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode). Is the documentation up to date?

Also, should `Send` fail would it make sense to expose the error code more explicitly? Right now it is hidden in a string message.

Reply:
We will start with the bare minimal API surface possible. We can add things like error codes (or possibly package-level error constants as is typical in Go), in the future depending on how developers use it. If we start with a large API surface now, it is going to be difficult to make changes without introducing breaking changes.

Reply:
I understand it, although it will make it much harder to process the errors programatically.
Moreover considering the other Admin SDKs (Python, Java, Node) will expose the error code, it would make sense to put Go SDK on par with the others.

Also, are the error codes in the docs obsolete?

Reply:
Sync'ed with the new error codes (documentation is up-to-date). We are still not handling some undocumented error codes, but we can deal with them later.